File: lib/cs/MsgPack/Internal/Reader.cs
using System;
using System.IO;
using System.Text;

namespace ILib.MsgPack
{
	public class Reader
	{
		byte[] m_Buf;
		int m_Pos;
		byte m_CurrentFormat;
		public byte CurrentFormat { get { return m_CurrentFormat; } }

		public Reader(byte[] buf, int pos = 0)
		{
			m_Buf = buf;
			m_Pos = pos;
		}

		public void Reset(byte[] buf, int pos = 0)
		{
			m_Buf = buf;
			m_Pos = pos;
		}

		public byte NextFormat()
		{
			return m_Buf[m_Pos];
		}

		public byte ReadFormat()
		{
			return m_CurrentFormat = m_Buf[m_Pos++];
		}

		public byte ReadPositiveFixInt()
		{
			return (byte)(m_CurrentFormat & 0x7f);
		}

		public byte ReadUInt8()
		{
			return m_Buf[m_Pos++];
		}

		public ushort ReadUInt16()
		{
			return (ushort)((m_Buf[m_Pos++] << 8) | m_Buf[m_Pos++]);
		}

		public uint ReadUInt32()
		{
			return ((uint)m_Buf[m_Pos++] << 24) | ((uint)m_Buf[m_Pos++] << 16) | ((uint)m_Buf[m_Pos++] << 8) | (uint)m_Buf[m_Pos++];
		}

		public ulong ReadUInt64()
		{
			return ((ulong)m_Buf[m_Pos++] << 56) | ((ulong)m_Buf[m_Pos++] << 48) | ((ulong)m_Buf[m_Pos++] << 40) | ((ulong)m_Buf[m_Pos++] << 32) | ((ulong)m_Buf[m_Pos++] << 24) | ((ulong)m_Buf[m_Pos++] << 16) | ((ulong)m_Buf[m_Pos++] << 8) | (ulong)m_Buf[m_Pos++];
		}

		public sbyte ReadNegativeFixInt()
		{
			return (sbyte)((m_CurrentFormat & 0x1f) - 0x20);
		}

		public sbyte ReadInt8()
		{
			return (sbyte)m_Buf[m_Pos++];
		}

		public short ReadInt16()
		{
			return (short)((m_Buf[m_Pos++] << 8) | m_Buf[m_Pos++]);
		}

		public int ReadInt32()
		{
			return (m_Buf[m_Pos++] << 24) | (m_Buf[m_Pos++] << 16) | (m_Buf[m_Pos++] << 8) | m_Buf[m_Pos++];
		}

		public long ReadInt64()
		{
			return ((long)m_Buf[m_Pos++] << 56) | ((long)m_Buf[m_Pos++] << 48) | ((long)m_Buf[m_Pos++] << 40) | ((long)m_Buf[m_Pos++] << 32) | ((long)m_Buf[m_Pos++] << 24) | ((long)m_Buf[m_Pos++] << 16) | ((long)m_Buf[m_Pos++] << 8) | (long)m_Buf[m_Pos++];
		}

		public float ReadFloat32()
		{
			m_Pos += 4;
			return FloatUnion.Read(m_Buf, m_Pos - 4);
		}

		public double ReadFloat64()
		{
			m_Pos += 8;
			return DoubleUnion.Read(m_Buf, m_Pos - 8);
		}

		public string ReadFixStr()
		{
			return ReadStringOfLength(m_CurrentFormat & 0x1f);
		}

		public string ReadStr8()
		{
			return ReadStringOfLength(ReadUInt8());
		}

		public string ReadStr16()
		{
			return ReadStringOfLength(ReadUInt16());
		}

		public string ReadStr32()
		{
			return ReadStringOfLength(Convert.ToInt32(ReadUInt32()));
		}

		public ArraySegment<byte> ReadBin8()
		{
			return ReadBytesOfLength(ReadUInt8());
		}

		public ArraySegment<byte> ReadBin16()
		{
			return ReadBytesOfLength(ReadUInt16());
		}

		public ArraySegment<byte> ReadBin32()
		{
			return ReadBytesOfLength(Convert.ToInt32(ReadUInt32()));
		}

		public int ReadArrayLength()
		{
			switch (m_CurrentFormat)
			{
				case Format.Nil: return 0;
				case Format.Array16: return ReadUInt16();
				case Format.Array32: return (int)ReadUInt32();
			}
			if (Format.IsFixArray(m_CurrentFormat))
			{
				return (byte)(m_CurrentFormat & 0xf);
			}
			throw new InvalidOperationException(string.Format("msgpack format invalid {0:X2}", m_CurrentFormat));
		}

		public int ReadMapLength()
		{
			switch (m_CurrentFormat)
			{
				case Format.Map16: return ReadUInt16();
				case Format.Map32: return (int)ReadUInt32();
			}
			if (Format.IsFixMap(m_CurrentFormat))
			{
				return (byte)(m_CurrentFormat & 0xf);
			}
			throw new InvalidOperationException(string.Format("msgpack format invalid {0:X2}", m_CurrentFormat));
		}

		public uint ReadExtLength()
		{
			switch (m_CurrentFormat)
			{
				case Format.FixExt1: return 1;
				case Format.FixExt2: return 2;
				case Format.FixExt4: return 4;
				case Format.FixExt8: return 8;
				case Format.FixExt16: return 16;
				case Format.Ext8: return ReadUInt8();
				case Format.Ext16: return ReadUInt16();
				case Format.Ext32: return ReadUInt32();
			}
			throw new InvalidOperationException(string.Format("msgpack format invalid {0:X2}", m_CurrentFormat));
		}

		public sbyte ReadExtType()
		{
			if (m_CurrentFormat == Format.UInt8)
			{
				return (sbyte)ReadUInt8();
			}
			else if (m_CurrentFormat == Format.Int8)
			{
				return (sbyte)ReadUInt8();
			}
			else if (Format.IsPositiveFixInt(m_CurrentFormat))
			{
				return (sbyte)ReadPositiveFixInt();
			}
			else if (Format.IsNegativeFixInt(m_CurrentFormat))
			{
				return ReadNegativeFixInt();
			}
			throw new InvalidOperationException(string.Format("msgpack format invalid {0:X2}", m_CurrentFormat));
		}

		public void Skip()
		{
			switch (ReadFormat())
			{
				case Format.Nil:
				case Format.False:
				case Format.True:
					return;
				case Format.UInt8:
				case Format.Int8:
					m_Pos += 1;
					return;
				case Format.UInt16:
				case Format.Int16:
					m_Pos += 2;
					return;
				case Format.UInt32:
				case Format.Int32:
				case Format.Float32:
					m_Pos += 4;
					return;
				case Format.UInt64:
				case Format.Int64:
				case Format.Float64:
					m_Pos += 8;
					return;
				case Format.Str8:
				case Format.Bin8:
					m_Pos += ReadUInt8();
					return;
				case Format.Str16:
				case Format.Bin16:
					m_Pos += ReadUInt16();
					return;
				case Format.Str32:
				case Format.Bin32:
					m_Pos += (int)ReadUInt32();
					return;
				case Format.FixExt1:
					m_Pos += 2;
					return;
				case Format.FixExt2:
					m_Pos += 3;
					return;
				case Format.FixExt4:
					m_Pos += 5;
					return;
				case Format.FixExt8:
					m_Pos += 9;
					return;
				case Format.FixExt16:
					m_Pos += 17;
					return;
				case Format.Ext8:
					m_Pos += 1 + ReadUInt8();
					return;
				case Format.Ext16:
					m_Pos += 1 + ReadUInt16();
					return;
				case Format.Ext32:
					m_Pos += 1 + (int)ReadUInt32();
					return;
			}
			if (Format.IsPositiveFixInt(m_CurrentFormat) || Format.IsNegativeFixInt(m_CurrentFormat))
			{
				return;
			}
			else if (Format.IsFixStr(m_CurrentFormat))
			{
				m_Pos += m_CurrentFormat & 0x1f;
			}
			else if (Format.IsArrayFamily(m_CurrentFormat))
			{
				for (int length = ReadArrayLength(); length > 0; length--)
				{
					Skip();
				}
			}
			else if (Format.IsMapFamily(m_CurrentFormat))
			{
				for (int length = ReadMapLength(); length > 0; length--)
				{
					Skip();
					Skip();
				}
			}
		}

		string ReadStringOfLength(int length)
		{
			m_Pos += length;
			return Encoding.UTF8.GetString(m_Buf, m_Pos - length, length);
		}

		internal ArraySegment<byte> ReadBytesOfLength(int length)
		{
			m_Pos += length;
			return new ArraySegment<byte>(m_Buf, m_Pos - length, length);
		}
	}
}

File: lib/golang/reader.go
package protopack

import (
	"bufio"
	"encoding/binary"
	"fmt"
	"io"
	"math"
)

type msgReader struct {
	r             *bufio.Reader
	currentFormat byte
}

func newMsgReader(reader io.Reader) *msgReader {
	return &msgReader{
		r: bufio.NewReader(reader),
	}
}

func (r *msgReader) Reset(reader io.Reader) {
	r.r = bufio.NewReader(reader)
}

func (r *msgReader) NextFormat() (byte, error) {
	buf, err := r.r.Peek(1)
	if err != nil {
		return 0, err
	}
	return buf[0], err
}

func (r *msgReader) ReadFormat() (byte, error) {
	format, err := r.r.ReadByte()
	r.currentFormat = format
	return format, err
}

func (r *msgReader) ReadPositiveFixInt() byte {
	return byte((r.currentFormat & 0x7f))
}

func (r *msgReader) ReadUInt8() (byte, error) {
	return r.r.ReadByte()
}

func (r *msgReader) ReadUInt16() (uint16, error) {
	buf := make([]byte, 2)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint16(buf), err
}

func (r *msgReader) ReadUInt32() (uint32, error) {
	buf := make([]byte, 4)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint32(buf), err
}

func (r *msgReader) ReadUInt64() (uint64, error) {
	buf := make([]byte, 8)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return binary.BigEndian.Uint64(buf), err
}

func (r *msgReader) ReadNegativeFixInt() int8 {
	return int8(((r.currentFormat & 0x1f) - 0x20))
}

func (r *msgReader) ReadInt8() (int8, error) {
	ret, err := r.r.ReadByte()
	if err != io.EOF && err != nil {
		return 0, err
	}
	return int8(ret), err
}

func (r *msgReader) ReadInt16() (int16, error) {
	buf := make([]byte, 2)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return int16(binary.BigEndian.Uint16(buf)), err
}

func (r *msgReader) ReadInt32() (int32, error) {
	buf := make([]byte, 4)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return int32(binary.BigEndian.Uint32(buf)), err
}

func (r *msgReader) ReadInt64() (int64, error) {
	buf := make([]byte, 8)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return int64(binary.BigEndian.Uint64(buf)), err
}

func (r *msgReader) ReadFloat32() (float32, error) {
	buf := make([]byte, 4)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return math.Float32frombits(binary.BigEndian.Uint32(buf)), err
}

func (r *msgReader) ReadFloat64() (float64, error) {
	buf := make([]byte, 8)
	_, err := r.r.Read(buf)
	if err != io.EOF && err != nil {
		return 0, err
	}
	return math.Float64frombits(binary.BigEndian.Uint64(buf)), err
}

func (r *msgReader) ReadFixStr() (string, error) {
	len := byte(r.currentFormat & 0x1f)
	ret := make([]byte, len)
	_, err := r.r.Read(ret)
	if err != io.EOF && err != nil {
		return "", err
	}
	return string(ret), err
}

func (r *msgReader) ReadStr8() (string, error) {
	len, err := r.ReadUInt8()
	if err != nil {
		return "", err
	}
	ret := make([]byte, len)
	_, err = r.r.Read(ret)
	if err != io.EOF && err != nil {
		return "", err
	}
	return string(ret), err
}

func (r *msgReader) ReadStr16() (string, error) {
	len, err := r.ReadUInt16()
	if err != nil {
		return "", err
	}
	buf := make([]byte, len)
	_, err = r.r.Read(buf)
	if err != io.EOF && err != nil {
		return "", err
	}
	return string(buf), err
}

func (r *msgReader) ReadStr32() (string, error) {
	len, err := r.ReadUInt32()
	if err != nil {
		return "", err
	}
	buf := make([]byte, len)
	_, err = r.r.Read(buf)
	if err != io.EOF && err != nil {
		return "", err
	}
	return string(buf), err
}

func (r *msgReader) ReadBin8() ([]byte, error) {
	len, err := r.ReadUInt8()
	if err != nil {
		return nil, err
	}
	ret := make([]byte, len)
	_, err = r.r.Read(ret)
	if err != io.EOF && err != nil {
		return nil, err
	}
	return ret, err
}

func (r *msgReader) ReadBin16() ([]byte, error) {
	len, err := r.ReadUInt16()
	if err != nil {
		return nil, err
	}
	buf := make([]byte, len)
	_, err = r.r.Read(buf)
	if err != io.EOF && err != nil {
		return nil, err
	}
	return buf, err
}

func (r *msgReader) ReadBin32() ([]byte, error) {
	len, err := r.ReadUInt32()
	if err != nil {
		return nil, err
	}
	buf := make([]byte, len)
	_, err = r.r.Read(buf)
	if err != io.EOF && err != nil {
		return nil, err
	}
	return buf, err
}

func (r *msgReader) ReadArrayLength() (uint, error) {
	switch r.currentFormat {
	case formatNil:
		return 0, nil
	case formatArray16:
		ret, err := r.ReadUInt16()
		return uint(ret), err
	case formatArray32:
		ret, err := r.ReadUInt32()
		return uint(ret), err
	}
	if isFixArray(r.currentFormat) {
		return uint(r.currentFormat & 0xf), nil
	}
	return 0, fmt.Errorf("msgpack format invalid %v", r.currentFormat)
}

func (r *msgReader) ReadMapLength() (uint, error) {
	switch r.currentFormat {
	case formatMap16:
		ret, err := r.ReadUInt16()
		return uint(ret), err
	case formatMap32:
		ret, err := r.ReadUInt32()
		return uint(ret), err
	}
	if isFixMap(r.currentFormat) {
		return uint(r.currentFormat & 0xf), nil
	}
	return 0, fmt.Errorf("msgpack format invalid %v", r.currentFormat)
}

func (r *msgReader) ReadExtLength() (uint, error) {
	switch r.currentFormat {
	case formatFixExt1:
		return 1, nil
	case formatFixExt2:
		return 2, nil
	case formatFixExt4:
		return 4, nil
	case formatFixExt8:
		return 8, nil
	case formatFixExt16:
		return 16, nil
	case formatExt8:
		ret, err := r.ReadUInt8()
		return uint(ret), err
	case formatExt16:
		ret, err := r.ReadUInt16()
		return uint(ret), err
	case formatExt32:
		ret, err := r.ReadUInt32()
		return uint(ret), err
	}
	return 0, fmt.Errorf("msgpack format invalid %v", r.currentFormat)
}

func (r *msgReader) ReadExtType() (int8, error) {
	if r.currentFormat == formatUInt8 {
		ret, err := r.ReadUInt8()
		return int8(ret), err
	} else if r.currentFormat == formatInt8 {
		ret, err := r.ReadUInt8()
		return int8(ret), err
	} else if isPositiveFixInt(r.currentFormat) {
		return int8(r.ReadPositiveFixInt()), nil
	} else if isNegativeFixInt(r.currentFormat) {
		return r.ReadNegativeFixInt(), nil
	}
	return 0, fmt.Errorf("msgpack format invalid %v", r.currentFormat)
}

func (r *msgReader) seek(n int) error {
	_, err := r.r.Discard(n)
	return err
}

func (r *msgReader) Skip() error {
	format, err := r.ReadFormat()
	if err != nil {
		return err
	}
	switch format {
	case formatNil:
	case formatFalse:
	case formatTrue:
		return nil
	case formatUInt8:
	case formatInt8:
		return r.seek(1)
	case formatUInt16:
	case formatInt16:
		return r.seek(2)
	case formatUInt32:
	case formatInt32:
	case formatFloat32:
		return r.seek(4)
	case formatUInt64:
	case formatInt64:
	case formatFloat64:
		return r.seek(8)
	case formatStr8:
	case formatBin8:
		len, err := r.ReadUInt8()
		if err != nil {
			return err
		}
		return r.seek(int(len))
	case formatStr16:
	case formatBin16:
		len, err := r.ReadUInt16()
		if err != nil {
			return err
		}
		return r.seek(int(len))
	case formatStr32:
	case formatBin32:
		len, err := r.ReadUInt32()
		if err != nil {
			return err
		}
		return r.seek(int(len))
	case formatFixExt1:
		return r.seek(2)
	case formatFixExt2:
		return r.seek(3)
	case formatFixExt4:
		return r.seek(5)
	case formatFixExt8:
		return r.seek(9)
	case formatFixExt16:
		return r.seek(17)
	case formatExt8:
		len, err := r.ReadUInt8()
		if err != nil {
			return err
		}
		return r.seek(int(len) + 1)
	case formatExt16:
		len, err := r.ReadUInt16()
		if err != nil {
			return err
		}
		return r.seek(int(len) + 1)
	case formatExt32:
		len, err := r.ReadUInt32()
		if err != nil {
			return err
		}
		return r.seek(int(len) + 1)
	}
	if isPositiveFixInt(r.currentFormat) || isNegativeFixInt(r.currentFormat) {
		return nil
	} else if isFixStr(r.currentFormat) {
		return r.seek(int(r.currentFormat & 0x1f))
	} else if isArrayFamily(r.currentFormat) {
		len, err := r.ReadArrayLength()
		if err != nil {
			return err
		}
		for i := 0; i < int(len); i++ {
			err = r.Skip()
			if err != nil {
				return err
			}
		}
	} else if isMapFamily(r.currentFormat) {
		len, err := r.ReadMapLength()
		if err != nil {
			return err
		}
		len *= 2
		for i := 0; i < int(len); i++ {
			err = r.Skip()
			if err != nil {
				return err
			}
		}
	}
	return nil
}
